### ISSUE TITLE
Function typedefs drop type information

### ISSUE THREAD

**Issue by stacksmith** [rpav/c2ffi#50]
Hello again.  I am generating a spec from some raspbian headers.  This header line:
`typedef void VCHI_MESSAGE_DRIVER_EVENT_CALLBACK_T( void *state );`
is translated to:
`{ "tag": "typedef", "ns": 0, "name": "VCHI_MESSAGE_DRIVER_EVENT_CALLBACK_T", "location": "raspbian/usr/include/interface/vchi/message_drivers/message.h:156:14", "type": { "tag": ":function" } }`
losing the void return type as well as the (void* state) parameter entirely.
I am using up-to-date 4.0 LLVM/CLANG this time:)

Any thoughts?

**Comment by rpav** [comment]
Yeah, at least in CL this isn't really necessary—it's just a typedef, and basically all you care about is it being a function-pointer type—so I never really bothered to add this information.  If you're doing something fancy like code completion or similar, it could probably be added without a lot of trouble or breaking compatibility.  I don't know if/when I'd have time to do this myself though.

**Comment by stacksmith** [comment]
Interestingly, I think it is **not a function pointer**, but a **function** stand in, describing a function.  This particular one is a problem for my FFI, as the function itself is never declared anywhere else - it is a callback function description.  So yes, the actual return type and parameters are absolutely necessary in this context.

https://stackoverflow.com/questions/3674200/what-does-a-typedef-with-parenthesis-like-typedef-int-fvoid-mean-is-it-a)

**Comment by rpav** [comment]
Well, you can't actually *use* them for anything.  It doesn't describe an *actual* function, just the "type" of a function.  In C/C++, this is used more or less exclusively for validating a function pointer assignment, i.e., as per the post, if you were to do

```c
typedef int (functype)(void);

void bar(void) { .. }

functype *fn = bar;
```

...you'd probably get at least a warning.  But in the context of `c2ffi`, this is irrelevant: `bar` would be the function whose declaration you care about, and if any structs etc had this as an element, it only matters (to autowrap, and CFFI last I checked) that this is a function pointer you assign... no actual type checking is done.

Additionally for the *declaration* case in that post, you still get full declarations for the actual functions.

In this case, it looks like a callback declaration type, which is probably used for passing "correct" function pointers around in C, but in CL it doesn't matter.  You need to explicitly declare your `defcallback`, and you don't need to cast the `callback` pointer to pass or assign it.  Structs members will just be `:pointer`.   So as above, even if the info was there, there's just no place to really apply it to anything.

**Comment by stacksmith** [comment]
Mostly agree.  In my case I've written (well, not completely :)a cl library to inspect and reason about ffis generated by c2ffi (and possibly generate various kinds of bindings).  With it I expect to be able to generate callback function types  based on information collected, and oddly enough, the typedef above is the _only_ description of the function.  In any event, it seems that c2ffi's mission is to preserve any useful information for the next level to handle, and my argument here, is that the function type information is indeed useful.

My goal is to gather a complete database of all types as seen by a particular incantation of c2ffi, be able to inspect it, and interactively select a useful subset (a tree-shake of sorts).  It would be nice to have a _complete_ record of types, and the typedef may be the only such record (proven by at least this one case).

It is not a really important issue, but I'd love to see it fixed when you can.

**Comment by rpav** [comment]
Yeah if you're doing something more fancy with the data, it could be useful.  Will tag as feature.

**Comment by stacksmith** [comment]
Thanks.
